File: node/m10-sdk/src/types/role_binding.ts
import type { m10 } from "../../protobufs";
import { M10Error } from "../error";
import { DocumentId } from "../ids/document_id";
import { PublicKey } from "../ids/public_key";
import { unwrap } from "../utils";


export interface IRoleBinding {
    id: DocumentId;
    owner: PublicKey;
    name: string;
    roleId: DocumentId;
    subjects: PublicKey[];
    expressions: m10.sdk.IExpression[];
    isUniversal: boolean;
}

export class RoleBinding implements IRoleBinding {
    public id: DocumentId;
    public owner: PublicKey;
    public name: string;
    public roleId: DocumentId;
    public subjects: PublicKey[];
    public expressions: m10.sdk.IExpression[];
    public isUniversal: boolean;


    public constructor(properties: IRoleBinding) {
        this.id = properties.id;
        this.owner = properties.owner;
        this.name = properties.name;
        this.roleId = properties.roleId;
        this.subjects = properties.subjects;
        this.expressions = properties.expressions;
        this.isUniversal = properties.isUniversal;
    }


    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFrom(roleBinding: m10.sdk.IRoleBinding): RoleBinding {
        return new RoleBinding({
            id: DocumentId.fromUint8Array("role-binding", unwrap(roleBinding.id, M10Error.InvalidTransaction())),
            owner: PublicKey.fromUint8Array(unwrap(roleBinding.owner, M10Error.InvalidTransaction())),
            name: unwrap(roleBinding.name, M10Error.InvalidTransaction()),
            roleId: DocumentId.fromUint8Array("role", unwrap(roleBinding.role, M10Error.InvalidTransaction())),
            subjects: unwrap(roleBinding.subjects, M10Error.InvalidTransaction()).map(PublicKey.fromUint8Array),
            expressions: unwrap(roleBinding.expressions, M10Error.InvalidTransaction()),
            isUniversal: unwrap(roleBinding.isUniversal, M10Error.InvalidTransaction()),
        });
    }
}

File: rust/sdk/src/types/role_binding.rs
use crate::collections::ResourceId;
use crate::error::M10Error;
use crate::types::PublicKey;
use m10_protos::sdk;
use m10_protos::sdk::Expression;
use serde::Serialize;

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[cfg_attr(
    feature = "format",
    display("RoleBinding{{ id={id} owner={owner} name={name} role_id={role_id} is_universal={is_universal} created_at={created_at} updated_at={updated_at} created_by={created_by} description={description} expires_at={expires_at:?} labels={labels:?} }}")
)]
#[derive(Clone, Debug, Serialize)]
pub struct RoleBinding {
    pub id: ResourceId,
    pub owner: PublicKey,
    pub name: String,
    pub role_id: ResourceId,
    pub subjects: Vec<bytes::Bytes>,
    pub expressions: Vec<Expression>,
    pub is_universal: bool,
    pub created_at: u64,
    pub updated_at: u64,
    pub created_by: PublicKey,
    pub description: String,
    pub labels: std::collections::HashMap<String, String>,
    pub expires_at: Option<u64>,
}

impl TryFrom<sdk::RoleBinding> for RoleBinding {
    type Error = M10Error;

    fn try_from(role_binding: sdk::RoleBinding) -> Result<Self, Self::Error> {
        Ok(Self {
            id: ResourceId::try_from(role_binding.id.as_ref())?,
            owner: PublicKey(role_binding.owner.to_vec()),
            name: role_binding.name,
            role_id: ResourceId::try_from(role_binding.role.as_ref())?,
            subjects: role_binding.subjects,
            expressions: role_binding.expressions,
            is_universal: role_binding.is_universal,
            created_at: role_binding.created_at,
            updated_at: role_binding.updated_at,
            created_by: PublicKey(role_binding.created_by.to_vec()),
            description: role_binding.description,
            labels: role_binding.labels,
            expires_at: role_binding.expires_at,
        })
    }
}
